### PR TITLE
Index access for calldata and byte arrays.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libsolidity/CompilerUtils.cpp [hunk 1]
```diff
@@ -41,18 +41,22 @@ unsigned CompilerUtils::loadFromMemory(unsigned _offset, Type const& _type,
 	return loadFromMemoryHelper(_type, _fromCalldata, _padToWordBoundaries);
 }
 
-void CompilerUtils::loadFromMemoryDynamic(Type const& _type, bool _fromCalldata, bool _padToWordBoundaries)
+void CompilerUtils::loadFromMemoryDynamic(
+	Type const& _type, bool _fromCalldata, bool _padToWordBoundaries, bool _keepUpdatedMemoryOffset)
```
Comment (LefterisJP) [inline libsolidity/CompilerUtils.cpp]: In my latest PR had a lot of back and forth with Gavin regarding how breaking a function to multiple lines should be. Here is his comment that shed a light on it. https://github.com/ethereum/cpp-ethereum/pull/1169#issuecomment-76964724

So I suppose if you really want to break it you should add each argument to its own line.

File: libsolidity/CompilerUtils.h [hunk 2]
```diff
@@ -46,7 +46,8 @@ class CompilerUtils
 	/// Dynamic version of @see loadFromMemory, expects the memory offset on the stack.
 	/// Stack pre: memory_offset
 	/// Stack post: value... (memory_offset+length)
-	void loadFromMemoryDynamic(Type const& _type, bool _fromCalldata = false, bool _padToWordBoundaries = true);
+	void loadFromMemoryDynamic(Type const& _type, bool _fromCalldata = false,
```
Comment (LefterisJP) [inline libsolidity/CompilerUtils.h]: Same thing here. It should be:

```
void loadFromMemoryDynamic(
	Type const& _type,
	 bool _fromCalldata = false,
	bool _padToWordBoundaries = true,
	bool _keepUpdatedMemoryOffset = true
);
```

File: libsolidity/LValue.cpp [hunk 3]
```diff
@@ -232,6 +232,64 @@ void StorageItem::setToZero(SourceLocation const&, bool _removeReference) const
...
+	else
+		m_context << eth::Instruction::DUP2 << eth::Instruction::SLOAD
+			<< eth::Instruction::DUP2 << eth::Instruction::BYTE;
+}
+
+void StorageByteArrayElement::storeValue(Type const&, SourceLocation const&, bool _move) const
+{
+	//@todo optimize this
+
+	// stack: value ref bytenr
```
Comment (LefterisJP) [inline libsolidity/LValue.cpp]: Maybe add some `static const int` or `#define` here for all those constants? Of course if you plan to change/optimize soon, then it won't be needed.

File: libsolidity/LValue.cpp [hunk 4]
```diff
@@ -232,6 +232,64 @@ void StorageItem::setToZero(SourceLocation const&, bool _removeReference) const
...
+	m_context << u256(31) << eth::Instruction::SUB << u256(0x100) << eth::Instruction::EXP;
+	// stack: value ref (1<<(8*(31-bytenr)))
+	m_context << eth::Instruction::DUP2 << eth::Instruction::SLOAD;
+	// stack: value ref (1<<(8*(31-bytenr))) old_full_value
```
Comment (LefterisJP) [inline libsolidity/LValue.cpp]: in all those comments, is `bytenr` number of bytes? Wouldn't it be more understandable as `bytenum` or `byte_num`?

File: libsolidity/LValue.h [hunk 5]
```diff
@@ -114,6 +113,23 @@ class StorageItem: public LValue
...
+ * Stack: <storage_ref> <byte_number>
+ */
+class StorageByteArrayElement: public LValue
+{
+public:
+	/// Constructs the LValue and assumes that the storage reference is already on the stack.
+	StorageByteArrayElement(CompilerContext& _compilerContext);
+	virtual unsigned sizeOnStack() const override { return 2; }
+	virtual void retrieveValue(SourceLocation const& _location, bool _remove = false) const override;
+	virtual void storeValue(Type const& _sourceType,
```
Comment (LefterisJP) [inline libsolidity/LValue.h]: Same thing I mentioned above about breaking a function into multiple arguments each on its own line.

File: libsolidity/LValue.h [hunk 6]
```diff
@@ -114,6 +113,23 @@ class StorageItem: public LValue
...
+		SourceLocation const& _location = SourceLocation(), bool _move = false) const override;
+	virtual void setToZero(
```
Comment (LefterisJP) [inline libsolidity/LValue.h]: here too

File: libsolidity/LValue.cpp (hunk 3 above)
Comment (chriseth) [inline libsolidity/LValue.cpp]: I think naming them does not help in understanding this.

**Pull request merged**